cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,455 @@
+  * Q: How to make missing certificate errors super obvious for debugging? We're attempting to route around this problem by having Pod startup blocked until the SPIRE server has completed issuing certificates for the identity given to any specific Pod, the same as we already do for Identity creation.
+
+##### CiliumNetworkPolicy creation
+Parse CNPs for the new authentication field. When this is present, as part of policy plumbing, set a new field in the policymap entries that indicates that authentication is required for connectivity to this peer.
+This process will be skipped if the `mesh-auth-mode` config directive is not set.
+
+##### Triggered by datapath drop
+* When the datapath generates a datapath drop message with the reason that "authentication is required" for this peer, look up the corresponding node and establish an mTLS connection to the peer node. Connects to the port outlined in the Startup section above.
+  * Use the Cilium Identity to look up TLS details.
+  * Special case: peer node may be the same node as the current node. In this case, the Node can connect to its own serving port, and perform the handshake, just like a remote node.
+  * Validate certificate of the peer. Note that since we are also planning to handle re-key events by pre-emptively re-authenticating, this should not occur very often.
+    * Q: How to make invalid certificate errors super obvious for debugging?
+      * Logs
+      * Metrics
+
+##### Authenticated session established with peer
+
+* Update datapath to inform it that this peer is authenticated (see Data Plane for details)
+* The destination node is responsible for checking that a Pod matching the destination Identity is running on itself, and only allowing the authentication to succeed if it is.
+
+##### Endpoint deletion
+
+* Node with endpoint must clear out Auth table to remove any state that implies authentication occurred for connections with this endpoint (This should already exist)
+* Do all other nodes need to clear out their corresponding state also?
+  * Thinking of IP recycling case, e.g. in ENI mode an IP could move from one node to another.
+  * Trigger on CiliumEndpoint deletion even for pods on remote nodes, delete authentication state corresponding to that peer. Including userspace authentication table entries, datapath CT/map entries.
```

Comment:
Note that Cilium doesn't subscribe to `Pod` events for other nodes. However, each node _should_ receive `CiliumEndpoint` or `CiliumEndpointSlice` delete events (or equivalent etcd/kvstore events) which could be used for this purpose. CES may delay some period longer than CEP but I think that's a general concern for CES in this area, not specific to this CFP.

Reply:
I moved the mention to `CiliumEndpoints` instead, thanks!